### PR TITLE
	gnome.py: compile gschemas to $prefix/share/glib-2.0/schemas/

### PR DESCRIPTION
**Pull request by csoriano89**
There's probably a reason why it was not like this... but I couldn't figure out, so this is what I came up with.

### CODE REVIEW

File: mesonbuild/modules/gnome.py [hunk 1]
```diff
@@ -338,11 +338,9 @@ def generate_gir(self, state, args, kwargs):
...
 
     def compile_schemas(self, state, args, kwargs):
-        if len(args) != 0:
-            raise MesonException('Compile_schemas does not take positional arguments.')
-        srcdir = os.path.join(state.build_to_src, state.subdir)
-        outdir = state.subdir
-        cmd = ['glib-compile-schemas', '--targetdir', outdir, srcdir]
+        srcdir = os.path.join(state.build_to_src, state.subdir, args[0])
+        outdir = os.path.join(state.environment.get_prefix(), 'share/glib-2.0/schemas')
+        cmd = ['glib-compile-schemas', srcdir, '--targetdir', outdir]
```
Comment (ignatenkobrain) [inline mesonbuild/modules/gnome.py]: will it install it immediately to PREFIX?

Comment (csoriano89) [inline mesonbuild/modules/gnome.py]: yes, should it be to some local directory and then install_dir to PREFIX?

**Comment by TingPing** [comment]
I believe that it is only intended for when running things locally, there are separate commands to run things on install.

**Comment by csoriano89** [comment]
What's the solution for generating and installing gschemas then? I couldn't find a real example, apart of the one in the meson repo which doesn't work for a real situation.
I believe we could have a glib_install_schemas? or should that be done with install_data (the_compiled_schema)?

**Comment by TingPing** [comment]
@csoriano89 `meson.add_install_script()` then your script would call `glib-compile-schemas ...`

**Comment by csoriano89** [comment]
Thanks for the answer TingPing, but I mean a proper solution using meson support for GNOME, not a handmade script.

**Comment by ignatenkobrain** [comment]
@csoriano89 basically doing glib-compile-schemas is work for package manager, but we wanted some better integration with meson itself in #837

**Comment by csoriano89** [comment]
but... compiling the schemas is necessary for a regular application to work. How is that an only package manager work? In autotools we have it when installing.

**Comment by ignatenkobrain** [comment]
that's basically what we do now. We compile it into builddir.

**Comment by ignatenkobrain** [comment]
@csoriano89 
https://developer.gnome.org/gio/stable/running-gio-apps.html

```
GSETTINGS_SCHEMA_DIR.  This variable can be set to the name of a directory that is considered in addition to the glib-2.0/schemas subdirectories of the XDG system data dirs when looking for compiled schemas for GSettings. ```

**Comment by ignatenkobrain** [comment]
we must not modify user's system when building project

**Comment by csoriano89** [comment]
I think I didn't explain well. In autotools you can define the prefix and it will install the schemas and the application will work. Here I don't see how to do it with meson support.
> we must not modify user's system when building project
Yes, just when installing is fine.

**Comment by ignatenkobrain** [comment]
@csoriano89 right, let's close this in favor of #837 and add some note there.

**Comment by ignatenkobrain** [comment]
@csoriano89 
```
compiled = gnome.compile_schemas()
install_data('com.github.meson.gschema.xml',
install_dir : 'share/glib-2.0/schemas')
```

meanwhile.

**Comment by TingPing** [comment]
@csoriano89 The 'problem' is we don't always want to generate data files on install, generally users don't directly install software and packagers do so this doesn't apply to them and would only get in their way. So we need to standardize some method of detecting if this should be done or a configure flag to do it, etc.

**Comment by ignatenkobrain** [comment]
@TingPing I don't see what's the problem with compiling all schemas in system. It could take longer than just 1 schema, but who cares.

**Comment by TingPing** [comment]
@ignatenkobrain Sorry I just mis-read your reply, I thought you suggested compiling it then installing the compiled file, that wouldn't work. `gnome.compile_schemas()` really only is useful for running locally for tests or something.

**Comment by ignatenkobrain** [comment]
@TingPing definitely that would not work ;)

**Comment by csoriano89** [comment]
I was thinking in precisely compiling the schemas in the buildir, then install the compiled file in the installdir. That wouldn't work?

**Comment by csoriano89** [comment]
ignatenkobrain,  
install_data('com.github.meson.gschema.xml',
install_dir : 'share/glib-2.0/schemas')
this  doesn't work, it's needed to call glib-compile-schemas after that. Which is what I'm asking for to do automatically when installing, or compile it on building and then install the compiled result

**Comment by TingPing** [comment]
> I was thinking in precisely compiling the schemas in the buildir, then install the compiled file in the installdir. That wouldn't work?

No, the reason all of these tasks are post-install is because it combines many files into a single one and you require the full installed list to work.

**Comment by ignatenkobrain** [comment]
@csoriano89
> this doesn't work, it's needed to call glib-compile-schemas after that. Which is what I'm asking for to do automatically when installing, or compile it on building and then install the compiled result

That's work of package manager (and meson, but only when you are installing something locally). So basically this is something what we want to do in #837). Let's say post-install actions.

**Comment by csoriano89** [comment]
I see, everything clear now, thanks.

**Pull request closed without merging**